Allow mac address to be optionally defined (#63)

njk464 opened the pull request.

ddelnano commented:
Hi @njk464! Thanks for raising this feature and giving the implementation a shot.

Can you show your work for how you tested this? We will need to make sure the acceptance tests pass (and probably add a new one for this case). I can help with that since the test suite is hard coded for parts of my deployment at the moment.

njk464 commented:
Hi @ddelnano. I just tested it manually on our xenorchestra 5.46.0 instance with a vm resource for the cases when the `mac_address` field is defined, not defined, and defined incorrectly ("00:00:00:00:00:0H". In all cases it behaved as expected (assigning the defined mac address to the VIF, auto generating a mac address, and throwing an error). I could probably add something before the vm resource is created to verify the mac address if you want. Right now the error that's shown when the mac address is not defined is

```
Error: jsonrpc2: code -32000 message: unknown error from the peer

  on main.tf line 53, in resource "xenorchestra_vm" "bar":
  53: resource "xenorchestra_vm" "bar" 
```

As for making acceptance tests It looks like that would be in `client/vm_test.go`? But I don't even know where to start with making a test since right now the

ddelnano commented on xoa/resource_xenorchestra_vm.go:
```diff
@@ -183,13 +183,16 @@ func resourceVmCreate(d *schema.ResourceData, m interface{}) error {
-		network_ids = append(network_ids, net["network_id"].(string))
+		network_maps = append(network_maps, map[string]string{
+			"network_id": net["network_id"].(string),
+			"macAddress": net["mac_address"].(string),
```
Since the rest of these keys are snake case I would prefer if we made this `mac_address`.

ddelnano commented on xoa/resource_xenorchestra_vm.go:
```diff
@@ -127,7 +127,7 @@ func resourceRecord() *schema.Resource {
 						"mac_address": &schema.Schema{
 							Type:     schema.TypeString,
-							Computed: true,
+							Optional: true,
```
We should add `ValidateFunc` here so that we can detect malformed mac addresses. The `jsonrpc2: code -32000 message: unknown error from the peer` error messages are difficult to interpret for most users. The validation should make it clear that the user supplied invalid input instead of them thinking the provider or XO api is not working.

ddelnano reviewed (changes requested):
I added an example of the acceptance test on [this branch](https://github.com/terra-farm/terraform-provider-xenorchestra/commit/6cdeec1737e36ac11bef0d6e01b3dd93b3caad88#diff-7ef95e5de9287a3d6051a21be0a66bb7). It passes which shows your code change does what its supposed to do 🙌 .

Unfortunately it seems to break some existing tests. Specifically the vm creation test. The plan doesn't seem to converge to a noop. For some reason the provider "loses" the mac address after creation.

```
ddelnano@ddelnano-desktop:~/go/src/github.com/ddelnano/terraform-provider-xenorchestra$ TF_ACC=1 TF_LOG=DEBUG  ~/go/bin/go1.14.7 test ./... -v -run TestAccXenorchestraVm_create

...
...
[DEBUG] Setting network via f6:f7:2a:33:b0:4d-d225cf00-36f8-e6d6-6a29-02636d4de56b
[DEBUG] Setting network via f6:f7:2a:33:b0:4d-d225cf00-36f8-e6d6-6a29-02636d4de56b

# The next line the mac address disappears ....

[DEBUG] Setting network via -d225cf00-36f8-e6d6-6a29-02636d4de56b
2020/09/20 21:35:07 [DEBUG] provider has no plugin.Client

    testing.go:684: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: xenorchestra_vm.bar
          auto_poweron:          "false" => "false"
          cloud_config:          "template" => "template"
          core_os:               "false" => "false"
          cpu_cap:               "0" => "0"
          cpu_weight:            "0" => "0"
          cpus:                  "1" => "1"
          disk.#:                "1" => "1"
          disk.0.name_label:     "xo provider root" => "xo provider root"
          disk.0.size:           "10000000000" => "10000000000"
          disk.0.sr_id:          "7f469400-4a2b-5624-cf62-61e522e50ea1" => "7f469400-4a2b-5624-cf62-61e522e50ea1"
          id:                    "4190cecd-4e0b-234e-eba3-224120872167" => "4190cecd-4e0b-234e-eba3-224120872167"
          memory_max:            "256000000" => "256000000"
          name_description:      "description" => "description"
          name_label:            "Terraform testing" => "Terraform testing"
          network.#:             "1" => "1"
          network.0.attached:    "true" => "<computed>"
          network.0.device:      "0" => "<computed>"
          network.0.mac_address: "f6:f7:2a:33:b0:4d" => ""
          network.0.network_id:  "d225cf00-36f8-e6d6-6a29-02636d4de56b" => "d225cf00-36f8-e6d6-6a29-02636d4de56b"
          template:              "17edcc45-0663-6bae-375f-1481056800eb" => "17edcc45-0663-6bae-375f-1481056800eb"
```

I haven't ran all the vm tests since they are slow but this issue will need to be worked out before we can merge it.

ddelnano commented:
The issue with the test I identified above is because the property is no longer computed. Terraform thinks that on resource refresh the `mac_address` is an empty string (because the vm resource code doesn't include a `mac_address` attribute and it's optional) but the internals of the provider is treating it like a computed property (creating it behind the scenes and storing it in the state).

I'm sure there is a pattern to deal with this but I don't know off the top of my head.

njk464 commented:
Hey @ddelnano. I added the changes you requested, added the test cases you created, and updated the documentation. I'm not able to run the tests on my setup so please let me know if any other tests fail.

ddelnano commented:
Since I need to run the tests and I need a little patch on top of what you have (the test code doesn't have a function it needs) I'm going to close this in favor of #65. I cherry picked your commits and the tests are passing.

Thanks for this contribution and sorry the tests aren't approachable yet! I plan to make them easy to run from a clean install soon

njk464 commented:
Thanks for helping me make a pull request! This feature is really going to help my team out with automating our deployment. Would love to do some more development on this project once the tests are more approachable.
